src/datachain/lib/prefetcher.py

```diff
@@ -0,0 +1,65 @@
+from collections.abc import Generator, Iterable, Sequence
+from contextlib import nullcontext
+from functools import partial
+from typing import TYPE_CHECKING, Any, Optional, TypeVar, cast
+
+from fsspec.callbacks import DEFAULT_CALLBACK, Callback
+
+from datachain.asyn import AsyncMapper
+from datachain.cache import temporary_cache
+from datachain.lib.file import File
+
+if TYPE_CHECKING:
+    from contextlib import AbstractContextManager
+
+    from datachain.cache import DataChainCache as Cache
+    from datachain.catalog.catalog import Catalog
+
+
+T = TypeVar("T", bound=Sequence[Any])
+
+
+def noop(*args, **kwargs):
+    pass
+
+
+async def _prefetch_input(row: T, catalog: "Catalog", download_cb: Callback) -> T:
+    try:
+        callback = download_cb.increment_file_count
+    except AttributeError:
+        callback = noop
+
+    for obj in row:
+        if isinstance(obj, File):
+            prefetched = await obj._prefetch(catalog, download_cb)
+            if prefetched:
+                callback()
+    return row
+
+
+def clone_catalog_with_cache(catalog: "Catalog", cache: "Cache") -> "Catalog":
+    clone = catalog.copy()
+    clone.cache = cache
+    return clone
+
+
+def rows_prefetcher(
+    catalog: "Catalog",
+    rows: Iterable[T],
+    prefetch: int,
+    cache: Optional["Cache"] = None,
+    download_cb: Callback = DEFAULT_CALLBACK,
+) -> Generator[T, None, None]:
+    cache_ctx: AbstractContextManager[Cache]
+    if cache:
+        cache_ctx = nullcontext(cache)
+    else:
+        tmp_dir = catalog.cache.tmp_dir
+        assert tmp_dir
+        cache_ctx = temporary_cache(tmp_dir, prefix="prefetch-")
```

Comment:
We'll create a temporary cache if the `cache` is not passed or set to `None`. (The caller will only pass `cache` object if `cache=True` is set in settings).

In that case, `rows_prefetcher` is responsible for removing it at the end, which it does below.

To use a temporary cache, we have to clone a catalog and update `cache` to the temporary cache object that we just created. Not only that, we need to update `File` objects to use the new `catalog` instead of the old one.

Admittedly, the approach in this file is a bit messy.